lib/Dialect/AIG/Analysis/LongestPathAnalysis.cpp

```diff
@@ -172,47 +262,1100 @@ struct llvm::DenseMapInfo<Object> {
+                                  SmallVectorImpl<OpenPath> &results) {
+  return addLogicOp(op, bitPos, results);
+}
+
+LogicalResult LocalVisitor::visit(comb::AndOp op, size_t bitPos,
+                                  SmallVectorImpl<OpenPath> &results) {
+  return addLogicOp(op, bitPos, results);
+}
+
+LogicalResult LocalVisitor::visit(comb::OrOp op, size_t bitPos,
+                                  SmallVectorImpl<OpenPath> &results) {
+  return addLogicOp(op, bitPos, results);
+}
+
+LogicalResult LocalVisitor::visit(comb::XorOp op, size_t bitPos,
+                                  SmallVectorImpl<OpenPath> &results) {
+  return addLogicOp(op, bitPos, results);
+}
+
+LogicalResult LocalVisitor::visit(comb::MuxOp op, size_t bitPos,
+                                  SmallVectorImpl<OpenPath> &results) {
+  // Add a cost of 1 for the mux.
+  if (failed(addEdge(op.getCond(), 0, 1, results)) ||
+      failed(addEdge(op.getTrueValue(), bitPos, 1, results)) ||
+      failed(addEdge(op.getFalseValue(), bitPos, 1, results)))
+    return failure();
+  deduplicatePaths(results);
+  return success();
+}
+
+LogicalResult LocalVisitor::visit(comb::ExtractOp op, size_t bitPos,
+                                  SmallVectorImpl<OpenPath> &results) {
+  assert(getBitWidth(op.getInput()) > bitPos + op.getLowBit());
+  auto result = markEquivalent(op, bitPos, op.getInput(),
+                               bitPos + op.getLowBit(), results);
+  return result;
+}
+
+LogicalResult LocalVisitor::visit(comb::ReplicateOp op, size_t bitPos,
+                                  SmallVectorImpl<OpenPath> &results) {
+  return markEquivalent(op, bitPos, op.getInput(),
+                        bitPos % getBitWidth(op.getInput()), results);
+}
+
+LogicalResult LocalVisitor::markFanIn(Value value, size_t bitPos,
+                                      SmallVectorImpl<OpenPath> &results) {
+  results.emplace_back(circt::igraph::InstancePath(), value, bitPos);
+  return success();
+}
+
+LogicalResult LocalVisitor::visit(hw::WireOp op, size_t bitPos,
+                                  SmallVectorImpl<OpenPath> &results) {
+  return markEquivalent(op, bitPos, op.getInput(), bitPos, results);
+}
+
+LogicalResult LocalVisitor::visit(hw::InstanceOp op, size_t bitPos,
+                                  size_t resultNum,
+                                  SmallVectorImpl<OpenPath> &results) {
+  auto moduleName = op.getReferencedModuleNameAttr();
+  auto value = op->getResult(resultNum);
+
+  // If an instance graph is not available, we treat instance results as a
+  // fanIn.
+  if (!ctx->instanceGraph)
+    return markFanIn(value, bitPos, results);
+
+  // If an instance graph is available, then we can look up the module. It's an
+  // error when the module is not found.
+  auto *node = ctx->instanceGraph->lookup(moduleName);
+  if (!node || !node->getModule()) {
+    op.emitError() << "module " << moduleName << " not found";
+    return failure();
+  }
```

Comment:
Do you think we could turn this into an assert? I think the `hw.instance` verifier would already error if the module did not exist :thinking: